### ISSUE TITLE
Scheduled job executing more than 1 time

### ISSUE THREAD

**Issue by TallGibbs** [jupyter-server/jupyter-scheduler#476]
## Description
I scheduled a script to run at 12:00 PM on each weekday.  When the next scheduled time came around, the job appears to have executed 3 different times with their run times overlapping each other.  I noticed this because my script is setup to have an automatic email sent and I received 3 copies of the email all with different results from executing my script.  Then, I checked the jupyter lab "notebook job" tab and saw 3 results of "completed" for the same script.  However in my "notebook job definitions" area, the script was only scheduled 1 time.

I'm wondering if the environment selected to run the job in affected this.  I scheduled it to run in the "base" environment (anaconda3) but I actually built the script in "projects" environment.  Also, I had previously scheduled this job, but then made revisions to the script.  So I deleted the scheduled job, and then scheduled it again.  Could this have affected the results?

![image](https://github.com/jupyter-server/jupyter-scheduler/assets/100488887/6166f3e5-81ae-460b-81d2-6d2daf91a198)

Here is my revised scheduled job (sorry, this screenshot is taken from AFTER the original issue showed up and I made a new schedule before opening this ticket).
![image](https://github.com/jupyter-server/jupyter-scheduler/assets/100488887/e4a5d2f8-9552-4155-8da4-d10104c79229)

## Reproduce

I'm not actually sure how to reproduce the issue.  All I did was schedule the script to run and this issue happened.

## Expected behavior

I wanted the job to only run 1 time at the scheduled time.  My script includes an automatic email sent out and I expected to only see 1 email and 1 completed job in the "completed job" tab.

## Context
- Operating System and version:
Brower = Chrome
- Jupyter Server version: 2.10.0
- Jupyter Lab version: 4.0.8
- Jupyter Notebook version: 7.0.6

<details><summary>Troubleshoot Output</summary>
<pre>
Paste the output from running `jupyter troubleshoot` from the command line here.
You may want to sanitize the paths in the output.
</pre>
</details>

(I tried to paste my entire result but I got an error that body is too long).  I have manually trimmed some info out of this result.

(base) C:\>jupyter troubleshoot
$PATH:    

sys.path:
        C:\Users\XXXXX\AppData\Local\anaconda3\Scripts
        C:\Users\XXXXX\AppData\Local\anaconda3\python311.zip
        C:\Users\XXXXX\AppData\Local\anaconda3\DLLs
        C:\Users\XXXXX\AppData\Local\anaconda3\Lib
        C:\Users\XXXXX\AppData\Local\anaconda3
        C:\Users\XXXXX\AppData\Roaming\Python\Python311\site-packages
        C:\Users\XXXXX\AppData\Roaming\Python\Python311\site-packages\win32
        C:\Users\XXXXX\AppData\Roaming\Python\Python311\site-packages\win32\lib
        C:\Users\XXXXX\AppData\Roaming\Python\Python311\site-packages\Pythonwin
        C:\Users\XXXXX\AppData\Local\anaconda3\Lib\site-packages
        C:\Users\XXXXX\AppData\Local\anaconda3\Lib\site-packages\win32
        C:\Users\XXXXX\AppData\Local\anaconda3\Lib\site-packages\win32\lib
        C:\Users\XXXXX\AppData\Local\anaconda3\Lib\site-packages\Pythonwin

sys.executable:
        C:\Users\XXXXX\AppData\Local\anaconda3\python.exe

sys.version:
        3.11.5 | packaged by Anaconda, Inc. | (main, Sep 11 2023, 13:26:23) [MSC v.1916 64 bit (AMD64)]

platform.platform():
        Windows-10-10.0.19045-SP0

where jupyter:
        C:\Users\XXXXX\AppData\Local\anaconda3\Scripts\jupyter.exe
[Jupyter Troubleshoot Results.docx](https://github.com/jupyter-server/jupyter-scheduler/files/14052847/Jupyter.Troubleshoot.Results.docx)

**Comment by andrii-i** [comment]
Hi @TallGibbs. Thank you for opening this issue and proactively providing `jupyter troubleshoot` output, this is very helpful. 

Could you please send a screenshot of "Run on a schedule" section of the Create Job screen for the job that this happened/happening with? For the sake of reproduction I'd like to understand if you define the schedule via `Interval > Day` or `Interval > Custom schedule > cron expression` option.

<img width="529" alt="Screenshot 2024-01-25 at 11 18 26 AM" src="https://github.com/jupyter-server/jupyter-scheduler/assets/26686070/d402fedc-1ba2-42f1-99eb-5b66a29decaf">

**Comment by TallGibbs** [comment]
Here you go!

Creating the schedule:
![image](https://github.com/jupyter-server/jupyter-scheduler/assets/100488887/2d454858-e99e-4323-92b0-621ec6e22519)


Final output of the scheduled job from the "Job Definition" page:
![image](https://github.com/jupyter-server/jupyter-scheduler/assets/100488887/068c7aeb-7b04-411f-a4f3-ec1f9d8a585d)

**Comment by andrii-i** [comment]
Thank you for the details @TallGibbs. I was not able to reproduce this through clock manipulation, trying to reproduce "organically" by waiting 12 PM.

**Comment by TallGibbs** [comment]
@andrii-i Today the job completed two times.  However, yesterday afternoon I re-installed jupyter notebook version 6.5.4 through Anaconda Navigator GUI, then updated (back) to 7.0.6 and then deleted all my jobs in Jupyter lab.  Then I scheduled this job again and now it resulted in the job running twice:

![image](https://github.com/jupyter-server/jupyter-scheduler/assets/100488887/f66d912d-a1d1-4cac-ad08-51c368034cfc)


Here is my scheduled jobs:
![image](https://github.com/jupyter-server/jupyter-scheduler/assets/100488887/ed66320b-1edf-4177-8899-8d0bbd3fd7b0)

**Comment by TallGibbs** [comment]
I'm not going to touch anything today on the job scheduler and see what happens on Saturday, Sunday and Monday.  It should NOT run on the weekend and then should run again on Monday.

**Comment by JasonWeill** [comment]
I set up a daily job run on my Windows 10 machine that has run exactly once per day, every day, since I first configured it.

If you run `jupyter server list` on a command prompt in your same Conda environment, how many entries do you see under "Currently running servers"? If you have multiple Jupyter Server instances running, that might be related to this.

**Comment by TallGibbs** [comment]
I'm going to leave everything the same for tomorrow and see if it run 4 times.  Then, I'm going to make a new schedule that is NOT at 12:00 PM and see if that is related to the issue. @andrii-i

**Comment by JasonWeill** [comment]
Thanks! For what it's worth, my schedule was at 8:00 am, and the run time typically started within 10 seconds of the hour.

**Comment by TallGibbs** [comment]
> I set up a daily job run on my Windows 10 machine that has run exactly once per day, every day, since I first configured it.
> 
> If you run `jupyter server list` on a command prompt in your same Conda environment, how many entries do you see under "Currently running servers"? If you have multiple Jupyter Server instances running, that might be related to this.

I just ran that code on my activated environment and I see 5 servers running (I'm not sure if it is safe to post screenshot of full server addresses).  It may be confusing though because I also currently have Jupyter Notebook open with some code I am working on, while also running Jupyter Lab.

I'm guessing a better test it to close everything down, then open just an Anaconda Prompt and run that code (after activating my environment)?

**Comment by JasonWeill** [comment]
If you have multiple Jupyter Servers open with the Jupyter Scheduler server extension running, that might be related to the behavior you're seeing. If you keep only one such server running, does that fix the problem?

(This may be an enhancement opportunity to run jobs only once, even if multiple servers are running.)

**Comment by TallGibbs** [comment]
> If you have multiple Jupyter Servers open with the Jupyter Scheduler server extension running, that might be related to the behavior you're seeing. If you keep only one such server running, does that fix the problem?
> 
> (This may be an enhancement opportunity to run jobs only once, even if multiple servers are running.)

What do you think is most valuable for testing/documentation purposes?  I see 3 options I could do right now.

Option 1 = Do nothing, and see if script runs 1 additional time tomorrow at noon.
Option 2 = Close all jupyter servers, then proceed with essentially option 1 (let scheduled job run at Noon).
Option 3 = Close all jupyter servers, make a new scheduled job that runs today at 3:00 PM and see what happens

**Comment by andrii-i** [comment]
@TallGibbs please try option 2. Make sure you have only 1 instance of the jupyter server with `jupyter_scheduler` extension installed running and see if the duplication goes away. I would expect this to solve the problem. I tried the same schedule on windows with 1 instance of the jupyter server running running and for me every job is created only once:
![no_duplicates](https://github.com/jupyter-server/jupyter-scheduler/assets/26686070/f91eed78-c059-4aac-b728-704152aa046d)

**Comment by TallGibbs** [comment]
@andrii-i Ok, I physically closed all open tabs with anything Jupyter related and then opened an Anaconda Prompt and ran the command: jupyter server list

Here are the results from that:
![image](https://github.com/jupyter-server/jupyter-scheduler/assets/100488887/787cefb5-b984-4ca6-844c-b4a990f300f0)

It appears there are 3 servers still running even though I do not physically have any open browsers related to Jupyter.  So, to try and shut them down I ran this command: jupyter notebook stop 8888

![image](https://github.com/jupyter-server/jupyter-scheduler/assets/100488887/522addb7-204a-4625-a588-4ac1df092127)


I get the response "Could not stop server on 8888" within my Anaconda Prompt.  Same response for all 3 servers when I try the different ports.  It makes me think there is some conflicts between my currently installed Anaconda prompt and previous versions?  I had to uninstall and reinstall several times in the past few weeks due to some package conflicts.

**Comment by TallGibbs** [comment]
@andrii-i sorry for the additional post: it looks like I had to completely close Anaconda Navigator and now I show 0 servers running.

![image](https://github.com/jupyter-server/jupyter-scheduler/assets/100488887/a88c35df-6e4f-48ce-9473-72268036e281)

**Comment by TallGibbs** [comment]
Today's update: with all the servers shutdown the scheduled job did not run at all.  Is that the expected response?  Is there a way to execute a scheduled job without a server running?

Now that we verified no job ran, with no servers running, I'm going to schedule the file to run this afternoon and see what happens.

**Comment by andrii-i** [comment]
@TallGibbs thank you for testing and confirming that there is no duplication when single server is running. 

> Today's update: with all the servers shutdown the scheduled job did not run at all. Is that the expected response? Is there a way to execute a scheduled job without a server running?

It is an expected response. Jupyter Scheduler is composed of `jupyter_scheduler` server extension and @jupyterlab/scheduler lab extension. `jupyter_scheduler` server extension owns and manages job execution. So with no instances of server running there would be no jobs created, with multiple instances of the server running there would be multiple instances of the jobs created. 

> Looks good today, job ran successfully and only ran 1 time. Now, I will leave it untouched through tomorrow's scheduled job and verify it still only runs once.

Thanks again for testing this.

**Comment by dlqqq** [comment]
It seems like your issue has been resolved, so I'll close this for now. Feel free to continue discussion if needed however! 👋

**Comment by TallGibbs** [comment]
Thank you everyone for all the help.

I'm not sure this is directly the right place for this question, but thought I would ask anyways even if just for a redirect: is there a way for me to have a jupyter server session running without Anaconda Navigator open?  We are trying to find a replacement for Alteryx Server running scheduled jobs (at my company) and I'm hoping I can find a way with Jupyter where a schedule job can run even when my work computer is shutdown.  I'm not an expert in computer networks/servers (clearly), so I am kind of figuring this out as I go...

**Comment by JasonWeill** [comment]
I run JupyterLab by running `jupyter lab` from an Anaconda-enabled command prompt, and I haven't run Anaconda Navigator at all. You'll need to have a server running all the time, ideally something running in a data center or in a hosted environment (disclosure: I work for AWS), that stays up so that jobs get rerun.

**Comment by TallGibbs** [comment]
Another follow-up closely related to this issue.  Is this normal that using the command "jupyter server stop 8888" gives back the response "Could not stop server on 8888" but then when you run "jupyter server list" the server on port 8888 no longer shows running?  Seems like a flaw in the code when it tells me "could not stop server on 8888."

Here is my screenshot of this exact sequence:
![image](https://github.com/jupyter-server/jupyter-scheduler/assets/100488887/b992832b-57d9-43ee-b190-4eb20cb71c30)

**Comment by JasonWeill** [comment]
Jupyter Server has a separate issue queue: https://github.com/jupyter-server/jupyter_server/issues

**Comment by andrii-i** [comment]
@TallGibbs below are some good places to ask questions if you don't want to create an issue:
- Discourse https://discourse.jupyter.org
- Appropriate community meeting. Jupyter Server Contributing Hour happening on Thursday (tomorrow for me) is probably the most appropriate for your question. Please find calendar of community events with links to each event here: https://discourse.jupyter.org/t/jupyter-community-calendar/2485
